### ISSUE TITLE
[REQUEST] Support parallel testing

### ISSUE THREAD

**Issue by mrleeio** [fastruby/next_rails#98]
## Description

Splitting test files into multiple processes on continuous integration platforms has become common to reduce run time.

For example, on GitHub Actions, you might create a workflow which utilizes a [matrix strategy](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs). You can read a blog post on the subject [here](https://hackernoon.com/how-parallel-github-actions-jobs-can-run-your-rspec-tests-faster-in-ruby-on-rails-application-yfj34kw).

The compare mode for `next_rails` is too simplistic to account for this type of setup. Each process will create a different shitlist file which fails the diff comparison.

## Possible Implementation

Create a new comparison mode that checks file content.

Committed file:
```
  "./spec/file_one.rb": [
    "DEPRECATION WARNING: Using `return`, `break` or `throw` to exit a transaction block is deprecated without replacement. ...",
    "DEPRECATION WARNING: Using `return`, `break` or `throw` to exit a transaction block is deprecated without replacement. ..."
  ],
  "./spec/file_two.rb": [
    "DEPRECATION WARNING: Rendering actions with '.' in the name is deprecated: ...",
    "DEPRECATION WARNING: Rendering actions with '.' in the name is deprecated: ..."
  ]
```

Matrix Job One:
```
  "./spec/file_one.rb": [
    "DEPRECATION WARNING: Using `return`, `break` or `throw` to exit a transaction block is deprecated without replacement. ...",
    "DEPRECATION WARNING: Using `return`, `break` or `throw` to exit a transaction block is deprecated without replacement. ..."
  ]
```

Matrix Job Two:
```
  "./spec/file_two.rb": [
    "DEPRECATION WARNING: Rendering actions with '.' in the name is deprecated: ...",
    "DEPRECATION WARNING: Rendering actions with '.' in the name is deprecated: ..."
  ]
```

Both Matrix jobs fail when comparing using a diff. However, it would pass if we searched for that content in the committed file.

## Resources:

- [Matrix strategy](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs)
- [Parallel GitHub Actions for RSpec](https://hackernoon.com/how-parallel-github-actions-jobs-can-run-your-rspec-tests-faster-in-ruby-on-rails-application-yfj34kw)

**I will abide by the [code of conduct](https://github.com/fastruby/next_rails/blob/main/CODE_OF_CONDUCT.md)**

**Comment by mrleeio** [comment]
@etagwerker sorry I may have been unclear. I only use the default comparison mode.

```
DEPRECATION_TRACKER=compare rspec
```

When this is run on separate processes being compared against a single shitlist file, it will always fail. Spec distribution among processes is completely random.

**Comment by etagwerker** [comment]
> This command works wonderfully locally or when specs are run within a single process.

I think this library could support parallel testing using your local environment, but I think it'd be too complicated to make it work in CI. 

@JuanVqz just implemented a "shard merger" for deprecations that happen in separate jobs, but even that won't solve this problem. 

I might be missing something, so I won't close this issue until I hear back from @JuanVqz, @arielj, or @mrleeio

**Comment by JuanVqz** [comment]
@etagwerker I feel like the parallel support we recently added in #177 can cover this scenario if you can follow this in a CI workflow.

> [!WARNING]
> Make sure you pass the the correct ENV["TEST_ENV_NUMBER"] at node_index

Step 1: Run Save mode

```ruby
  DeprecationTracker.track_rspec(
    mode: ENV["DEPRECATION_TRACKER"],
    node_index: ENV["TEST_ENV_NUMBER"] # :save
  )
```

Step 2: After all processes finish, merge the shards into one file:

```ruby
DeprecationTracker.merge_shards("spec/support/deprecation_warning.shitlist.json")
```
or the command line

```bash
deprecations merge
```

Step 3: Compare mode

Switch to compare mode:

```ruby
  DeprecationTracker.track_rspec(
    shitlist_path: "spec/support/deprecation_warning.shitlist.json",
    mode: ENV["DEPRECATION_TRACKER"] # :compare
  )
```

I might be wrong about this fully solving your use case since we haven't tested it specifically with the parallel_tests gem, but the mechanism should work.

@mrleeio if you get a chance, please test it out and let us know how it goes. We're interested in knowing whether this works well for your setup.

**Comment by etagwerker** [comment]
@JuanVqz I agree, I think the new `DeprecationTracker::ShardMerger` in #177 does all it can to add support for parallel testing. 

Any other issues to improve support for parallel testing can be filed as new issues. Closing this one for now. 👍🏻
